src/Function/Metric/Bundles.agda

```diff
@@ -0,0 +1,139 @@
+
+------------------------------------------------------------------------
+-- PreMetric
+
+record PreMetric (a i ℓ₁ ℓ₂ ℓ₃ : Level)
+               : Set (suc (a ⊔ i ⊔ ℓ₁ ⊔ ℓ₂ ⊔ ℓ₃)) where
+  field
+    Carrier     : Set a
+    Image       : Set i
+    _≈_         : Rel Carrier ℓ₁
+    _≈ᵢ_        : Rel Image ℓ₂
+    _≤_         : Rel Image ℓ₃
+    0#          : Image
+    d           : DistanceFunction Carrier Image
+    isPreMetric : IsPreMetric _≈_ _≈ᵢ_ _≤_ 0# d
+
+  open IsPreMetric isPreMetric public
+
+  protoMetric : ProtoMetric a i ℓ₁ ℓ₂ ℓ₃
+  protoMetric = record
+    { isProtoMetric = isProtoMetric
+    }
+
+------------------------------------------------------------------------
+-- QuasiSemiMetric
+
+record QuasiSemiMetric (a i ℓ₁ ℓ₂ ℓ₃ : Level)
+                     : Set (suc (a ⊔ i ⊔ ℓ₁ ⊔ ℓ₂ ⊔ ℓ₃)) where
+  field
+    Carrier           : Set a
+    Image             : Set i
+    _≈_               : Rel Carrier ℓ₁
+    _≈ᵢ_              : Rel Image ℓ₂
+    _≤_               : Rel Image ℓ₃
+    0#                : Image
+    d                 : DistanceFunction Carrier Image
+    isQuasiSemiMetric : IsQuasiSemiMetric _≈_ _≈ᵢ_ _≤_ 0# d
+
+  open IsQuasiSemiMetric isQuasiSemiMetric public
+
+  preMetric : PreMetric a i ℓ₁ ℓ₂ ℓ₃
+  preMetric = record
+    { isPreMetric = isPreMetric
+    }
+
+  open PreMetric preMetric public
+    using (protoMetric)
+
+------------------------------------------------------------------------
+-- SemiMetric
+
+record SemiMetric (a i ℓ₁ ℓ₂ ℓ₃ : Level)
+                : Set (suc (a ⊔ i ⊔ ℓ₁ ⊔ ℓ₂ ⊔ ℓ₃)) where
+  field
+    Carrier      : Set a
+    Image        : Set i
+    _≈_          : Rel Carrier ℓ₁
+    _≈ᵢ_         : Rel Image ℓ₂
+    _≤_          : Rel Image ℓ₃
+    0#           : Image
+    d            : DistanceFunction Carrier Image
+    isSemiMetric : IsSemiMetric _≈_ _≈ᵢ_ _≤_ 0# d
+
+  open IsSemiMetric isSemiMetric public
+
+  quasiSemiMetric : QuasiSemiMetric a i ℓ₁ ℓ₂ ℓ₃
+  quasiSemiMetric = record
+    { isQuasiSemiMetric = isQuasiSemiMetric
+    }
+
+  open QuasiSemiMetric quasiSemiMetric public
+    using (protoMetric; preMetric)
+
+------------------------------------------------------------------------
+-- GeneralMetric
+
+-- Note that this package is not necessarily a metric in the classical
+-- sense as there is no way to ensure that the _∙_ operator really
+-- represents addition. See `Function.Metric.Nat` and
+-- `Function.Metric.Rational` for more specialised `Metric` and
+-- `UltraMetric` packages.
+
+record GeneralMetric (a i ℓ₁ ℓ₂ ℓ₃ : Level)
+                   : Set (suc (a ⊔ i ⊔ ℓ₁ ⊔ ℓ₂ ⊔ ℓ₃)) where
+  field
+    Carrier         : Set a
+    Image           : Set i
+    _≈_             : Rel Carrier ℓ₁
+    _≈ᵢ_            : Rel Image ℓ₂
+    _≤_             : Rel Image ℓ₃
+    0#              : Image
+    _∙_             : Op₂ Image
+    d               : DistanceFunction Carrier Image
+    isGeneralMetric : IsGeneralMetric _≈_ _≈ᵢ_ _≤_ 0# _∙_ d
+
```

Comment:
Should `_∙_` and `0#` form a monoid?
￼

Reply:
I don't think so...
- the `_∙_` operators for p-relaxed metrics and p-inframetrics are not associative. 
- 0 is not an identity element for ultrametrics over the reals, only over the positive reals.

Reply:
I had never heard of p-relaxed metrics and p-inframetrics until now.  Might it make sense to add a link to some reference?